silence/silence.go

```diff
@@ -780,7 +827,9 @@ func (s *Silences) SetBroadcast(f func([]byte)) {
 type state map[string]*pb.MeshSilence
 
 func (s state) merge(e *pb.MeshSilence, now time.Time) bool {
+	id := e.Silence.Id
 	if e.ExpiresAt.Before(now) {
+		delete(s, id) // Just in case it is still lingering.
```

Comment:
I am not sure this preserves the silence deletion logic. Wouldn't it delete all copies of a silence during cluster replication once a user expires it? Even though the silence is logically useless once it is expired, thereby safe to delete, the UI differentiates between expired and non-existing silences.

Reply:
I'm not an expert of this codebase, so I am not sure if I got this right. But I think you got confused here by a somewhat unfortunate naming. (Naming is hard, as we know…) There are essentially two kinds of expirations for silences:
1. The silence itself expires, i.e. it's `EndsAt` date has been reached, or it “was expired” by setting its `endsAt` date to “now”.
2. The silence expires from the storage, i.e. we don't want to handle it in the storage anymore, which is the case if its `ExpiresAt` date has been reached (which is saved in the wrapping `pb.MeshSilence` and not in the silence itself).

This line is about the second case, i.e. we are really not interested in this silence anymore at all. The reason why this `merge` method returns `false` in that case without doing anything is that we won't bother merging the silence into the storage if we are not interested in it anymore.

The problem this creates happens if this collides with the way we explicitly expire silences (the first case) in combination with a very short or zero retention time (something that should not happen in practice, but could be set): If we expire a silence (the first case), we set its `EndsAt` to “now”. This happens on a copy of the silence (for concurrency safety or something), and then we `merge` the modified, copied silence (which still has the same ID) back into the storage. But what we also do (and must do) is to set the `ExpiresAt` to “now” + retention time. In case that's a finite time, everything is fine, but if the retention time is zero, now `ExpiresAt` is also “now”. At the time we call `merge`, the “now” is already increased a bit, and thus, `merge` sees the silence as not only expired (the first case) but also as expired from the storage (the second case) and thus it returns false and doesn't do anything, which means that _the original silence we copied for the expiration is still in the storage in unchanged form_. And that's a bug.

Reply:
And I think I have found a better fix, thanks to the rubberducking above.
I think a cleaner solution is to use the same “now” for merging the silence as the “now” we used to set the silence as expired (the first case).
I'll modify the respective commit. (It requires the internal `setSilence` to take a “now” parameter similar to the `merge`.)

Reply:
I understand the reasoning for deleting the storage silence given that the to-be-merged silence is expired. Thanks for the detailed explanation.

>  The silence expires from the storage, i.e. we don't want to handle it in the storage anymore, which is the case if its `ExpiresAt` date has been reached (which is saved in the wrapping `pb.MeshSilence` and not in the silence itself).

@beorn7 I am not sure this is true. A silence that is expired should still be in the storage at least for some reasonable amount of time to give the user feedback that the silence is indeed expired.

Reply:
Yes, it "should". But if the user has set zero retention time, then that's what they get: zero retention. That's the current state of the code, and that's by design. We could enforce a minimum retention time, but that has nothing to do with this PR.

Please also note that I have removed the explicit deletion and just use a consistent "now" so that the bug that expired silences stay unchanged (provided a zero retention time has been configured) is fixed with this PR in a clean way.

Reply:
Right, understood. Thanks for that!

> Please also note that I have removed the explicit deletion and just use a consistent "now" so that the bug that expired silences stay unchanged (provided a zero retention time has been configured) is fixed with this PR in a clean way.

Yes, saw that. Looks good to me.